cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -1774,14 +1774,21 @@ public void checkCustomCaSecret(CertificateAuthority ca, Secret certSecret, Secr
-                                        X509Certificate cert = Ca.cert(certSecret, customCertSecret.getCertificate());
-                                        byte[] signature = MessageDigest.getInstance("SHA-256").digest(cert.getEncoded());
-                                        thumbprintPromise.complete(Base64.getEncoder().encodeToString(signature));
-                                    } catch (CertificateEncodingException | NoSuchAlgorithmException e) {
-                                        log.warn("{}: Failed to get certificate signature of {} from Secret {}.", reconciliation, customCertSecret.getCertificate(), customCertSecret.getSecretName());
-                                        thumbprintPromise.fail(new RuntimeException("Failed to get certificate signature of " + customCertSecret.getCertificate() + " from Secret " + certSecret.getMetadata().getName(), e));
-                                    }
+                                    if (!certSecret.getData().containsKey(customCertSecret.getCertificate())) {
+                                        log.warn("{}: Secret {} does not contain certificate under the key {}.", reconciliation, customCertSecret.getSecretName(), customCertSecret.getCertificate());
+                                        thumbprintPromise.fail("Secret " + customCertSecret.getSecretName() + " does not contain certificate under the key " + customCertSecret.getCertificate() + ".");
+                                    } else if (!certSecret.getData().containsKey(customCertSecret.getKey())) {
+                                        log.warn("{}: Secret {} does not contain custom certificate private key under the key {}.", reconciliation, customCertSecret.getSecretName(), customCertSecret.getKey());
+                                        thumbprintPromise.fail("Secret " + customCertSecret.getSecretName() + " does not contain custom certificate private key under the key " + customCertSecret.getKey() + ".");
+                                    } else
+                                        try {
+                                            X509Certificate cert = Ca.cert(certSecret, customCertSecret.getCertificate());
+                                            byte[] signature = MessageDigest.getInstance("SHA-256").digest(cert.getEncoded());
+                                            thumbprintPromise.complete(Base64.getEncoder().encodeToString(signature));
+                                        } catch (CertificateEncodingException | NoSuchAlgorithmException e) {
+                                            log.warn("{}: Failed to get certificate signature of {} from Secret {}.", reconciliation, customCertSecret.getCertificate(), customCertSecret.getSecretName());
+                                            thumbprintPromise.fail(new RuntimeException("Failed to get certificate signature of " + customCertSecret.getCertificate() + " from Secret " + certSecret.getMetadata().getName(), e));
```

Comment:
I don't think we should use `RuntimeException` (at least in the case of a `CertificateEncodingException`, which would indicate that the user didn't create the Secret with a cert in the correct format). Remember that the exception class name gets set in the resource status, so it ought to be something helpful. I guess we could use the existing `InvalidResourceException` (though I've always thought of that as being about the consumed CR, rather than other resources)? Or Maybe a new exception class. 

The above errors for missing cert and key should use the same exception type.

Reply:
I guess when the keys are missing it is really abotu invalid CR -> the keys listed there are wrong. When they exist but fail to decode it is a bit harder.